internal/node/node.go

```diff
@@ -0,0 +1,1354 @@
+package node
```

Comment:
parts of this needs to be in the public API. (https://github.com/cosmos/cosmos-sdk/blob/25ecec6c152aa6406da9da24151f412384a3d94b/server/start.go#L256).

Reply:
The whole point of this change in general is to break the API and provide better alternatives, so this was always going to be breaking and the SDK will have to make some amount of changes at some point. 


Having said that, the code you link really looks like the sdk is duplicating the DefaultNewNode function exactly, so I actually think that the new API is the correct one. The new `node.New` just calls `DefaultNewNode` which is this: https://github.com/tychoish/tendermint/blob/node-internalize/internal/node/node.go#L93

If the SDK needs to control an aspect of the node that *isn't* controllable via the config structure, then we should change the config object to add that functionally.

Reply:
Yes it is meant to be breaking but in one fell swoop means we can be breaking things in a way that renders some apps unusable or them needing to fork the entire project. I dont think this was the goal.

Reply:
Is it a fair assumption that people who are using tendermint are using go mod pinned to a specific branch? This, coupled with the speed that most users tend to upgrade, would seem to give us a lot of flexibility to make changes like this without breaking anyone's apps until they intend to upgrade.  

I think the implication of the fact that the `node` package exports and explicitly depends on the interfaces of packages including all of the reactors and `p2p` means that there actually isn't a way to do this that leaves a meaningfully usable interface in more steps. We could break peoples apps in smaller ways a few times over many more releases, but I don't think that's actually a better option. 

Do you have an idea of the kind of functionality we need to provide or preserve, and examples of the projects that you think are actually depending on (internal) tendermint functionality?

Reply:
I tend to think that it's better to break in one fell swoop than over many iterations. But seeing as we kind of wanted to keep 0.35 not so breaking do you think it may be better to do these changes in 0.36. Or alternatively do a similar thing to the p2p refactor where we keep a few of the old public interfaces that people kind of need and would find annoying for us to internalise until 0.36.

Reply:
> But seeing as we kind of wanted to keep 0.35 not so breaking do you think it may be better to do these changes in 0.36.

As far as I know, the go API is already expected to break somewhat in this release, and we're primarily going for an operationally simple release (e.g. no hard forks, or halts.) In a lot of ways, doing a change like this now means that we have 1-2 releases to fix anything that's really bad before 1.0, so we get a chance to stabalize.

We still may decide to push this further out, but I think it might be ok to do this now. 

> Or alternatively do a similar thing to the p2p refactor where we keep a few of the old public interfaces that people kind of need and would find annoying for us to internalise until 0.36.

I think there's less flexibility here than in the p2p because we're not expecting to keep two copies of the code active in the repo in the same way. In some ways, with API things, doing a bigger break can be easier because the compiler can enforce the rules and there isn't really the same kind of subtly. 

Let's talk about it more tomorrow/sometime.

Reply:
> I think there's less flexibility here than in the p2p because we're not expecting to keep two copies of the code active in the repo in the same way

I haven't looked into your implementation in any depth (going to do that now) but I would imagine the public API's that you might want to keep would just point directly to the internal ones so you're not really keeping two copies of everything.

> we're primarily going for an operationally simple release

yeah good point

Reply:
> I would imagine the public API's that you might want to keep would just point directly to the internal ones so you're not really keeping two copies of everything.

Totally, I think between this commit (potentially) landing and 0.35 we'll definitely move some things out of `internal` and into the new `node` package (and `types` etc,) to facilitate some this. 

--- 

On a related note, and this might be subtle, so I'll underscore it: the `Node` constructors now return a `service.Service` interface, which means users (even internal) won't be able to (easily) reach into the node instances itself which will make simplification and will ease using these types externally.